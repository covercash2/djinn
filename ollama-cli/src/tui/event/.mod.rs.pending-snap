{"run_id":"1732394054-558155000","line":100,"new":{"module_name":"ollama_cli__tui__event__tests","snapshot_name":"load_default_keymap","metadata":{"source":"ollama-cli/src/tui/event/mod.rs","assertion_line":100,"expression":"event_definitions","snapshot_kind":"text"},"snapshot":"EventDefinitions(\n    {\n        \"edit\": ActionMap(\n            {\n                Key {\n                    event: KeyEvent {\n                        code: Up,\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Up,\n                        },\n                    ),\n                }: Up,\n                Key {\n                    event: KeyEvent {\n                        code: Backspace,\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Backspace,\n                        },\n                    ),\n                }: Backspace,\n                Key {\n                    event: KeyEvent {\n                        code: Esc,\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Esc,\n                        },\n                    ),\n                }: Escape,\n                Key {\n                    event: KeyEvent {\n                        code: Down,\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Down,\n                        },\n                    ),\n                }: Down,\n                Key {\n                    event: KeyEvent {\n                        code: Left,\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Left,\n                        },\n                    ),\n                }: Left,\n                Key {\n                    event: KeyEvent {\n                        code: Enter,\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Enter,\n                        },\n                    ),\n                }: Enter,\n            },\n        ),\n        \"normal\": ActionMap(\n            {\n                Key {\n                    event: KeyEvent {\n                        code: Char(\n                            'q',\n                        ),\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Char(\n                                'q',\n                            ),\n                        },\n                    ),\n                }: Quit,\n                Key {\n                    event: KeyEvent {\n                        code: Up,\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Up,\n                        },\n                    ),\n                }: Up,\n                Key {\n                    event: KeyEvent {\n                        code: Char(\n                            'j',\n                        ),\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Char(\n                                'j',\n                            ),\n                        },\n                    ),\n                }: Down,\n                Key {\n                    event: KeyEvent {\n                        code: Esc,\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Esc,\n                        },\n                    ),\n                }: Escape,\n                Key {\n                    event: KeyEvent {\n                        code: Char(\n                            '$',\n                        ),\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Char(\n                                '$',\n                            ),\n                        },\n                    ),\n                }: End,\n                Key {\n                    event: KeyEvent {\n                        code: Down,\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Down,\n                        },\n                    ),\n                }: Down,\n                Key {\n                    event: KeyEvent {\n                        code: Char(\n                            'w',\n                        ),\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Char(\n                                'w',\n                            ),\n                        },\n                    ),\n                }: RightWord,\n                Key {\n                    event: KeyEvent {\n                        code: Char(\n                            'i',\n                        ),\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Char(\n                                'i',\n                            ),\n                        },\n                    ),\n                }: Edit,\n                Key {\n                    event: KeyEvent {\n                        code: Char(\n                            'b',\n                        ),\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Char(\n                                'b',\n                            ),\n                        },\n                    ),\n                }: LeftWord,\n                Key {\n                    event: KeyEvent {\n                        code: Backspace,\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Backspace,\n                        },\n                    ),\n                }: Backspace,\n                Key {\n                    event: KeyEvent {\n                        code: Enter,\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Enter,\n                        },\n                    ),\n                }: Enter,\n                Key {\n                    event: KeyEvent {\n                        code: Char(\n                            'h',\n                        ),\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Char(\n                                'h',\n                            ),\n                        },\n                    ),\n                }: Left,\n                Key {\n                    event: KeyEvent {\n                        code: Char(\n                            'k',\n                        ),\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Char(\n                                'k',\n                            ),\n                        },\n                    ),\n                }: Up,\n                Key {\n                    event: KeyEvent {\n                        code: Char(\n                            'r',\n                        ),\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Char(\n                                'r',\n                            ),\n                        },\n                    ),\n                }: Refresh,\n                Key {\n                    event: KeyEvent {\n                        code: Left,\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Left,\n                        },\n                    ),\n                }: Left,\n                Key {\n                    event: KeyEvent {\n                        code: Char(\n                            '0',\n                        ),\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Char(\n                                '0',\n                            ),\n                        },\n                    ),\n                }: Beginning,\n                Key {\n                    event: KeyEvent {\n                        code: Char(\n                            'a',\n                        ),\n                        modifiers: NONE,\n                        kind: Press,\n                        state: NONE,\n                    },\n                    node: Some(\n                        Node {\n                            modifiers: 0,\n                            key: Char(\n                                'a',\n                            ),\n                        },\n                    ),\n                }: Edit,\n            },\n        ),\n    },\n)"},"old":{"module_name":"ollama_cli__tui__event__tests","metadata":{"snapshot_kind":"text"},"snapshot":""}}
