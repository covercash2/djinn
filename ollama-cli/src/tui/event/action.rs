use std::collections::HashMap;

use keymap::KeyMap;
use serde::{Deserialize, Serialize};
use crate::{error::Result, tui::AppEvent};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ActionDefinition {
    action: Action,
    description: String,
}

impl From<Action> for ActionDefinition {
    fn from(action: Action) -> Self {
        ActionDefinition {
            action,
            description: String::new(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Deserialize)]
pub struct ActionMap(pub HashMap<KeyMap, Action>);

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize, strum::Display)]
#[serde(rename_all = "snake_case")]
pub enum Action {
    Beginning,
    End,
    Left,
    Right,
    Up,
    Down,
    Edit,
    LeftWord,
    RightWord,
    Refresh,
    Popup,
    Help,
    Enter,
    Escape,
    Backspace,
    Quit,
    #[serde(skip)]
    Unhandled(char),
    Nop,
}

pub trait ActionHandler {
    type Event;
    fn handle_action(&mut self, action: Action) -> Result<Option<AppEvent>>;
}
