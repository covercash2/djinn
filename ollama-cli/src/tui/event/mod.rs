use std::collections::HashMap;

use crossterm::event::{Event, KeyEvent};
use keymap::KeyMap;
use serde::{Deserialize, Serialize};

const DEFAULTS: &str = include_str!("../../../default_keymap.toml");

#[derive(Debug, PartialEq, Deserialize)]
pub struct EventProcessor {
    input_mode: InputMode,
    definitions: EventDefinitions,
}

impl Default for EventProcessor {
    fn default() -> Self {
        let definitions = toml::from_str(DEFAULTS).expect("should be able to load default keymaps");
        Self {
            input_mode: Default::default(),
            definitions,
        }
    }
}

impl EventProcessor {
    pub fn input_mode(&mut self, input_mode: InputMode) {
        self.input_mode = input_mode;
    }

    pub fn process(&self, event: Event) -> Action {
        match event {
            Event::FocusGained
            | Event::FocusLost
            | Event::Mouse(_)
            | Event::Paste(_)
            | Event::Resize(_, _) => Action::Nop,
            Event::Key(key_event) => self.process_key_event(key_event),
        }
    }

    pub fn process_key_event(&self, event: KeyEvent) -> Action {
        self.definitions
            .0
            .get(&self.input_mode)
            .and_then(|map| map.0.get(&KeyMap::from(event)))
            .copied()
            .unwrap_or(Action::Nop)
    }
}

#[derive(Default, Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum InputMode {
    #[default]
    Normal,
    Edit,
}

#[derive(Debug, PartialEq, Deserialize)]
pub struct EventDefinitions(HashMap<InputMode, ActionMap>);

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ActionDefinition {
    action: Action,
    description: String,
}

impl From<Action> for ActionDefinition {
    fn from(action: Action) -> Self {
        ActionDefinition {
            action,
            description: String::new(),
        }
    }
}

#[derive(Debug, PartialEq, Deserialize)]
pub struct ActionMap(HashMap<KeyMap, Action>);

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Action {
    Beginning,
    End,
    Left,
    Right,
    Up,
    Down,
    Edit,
    LeftWord,
    RightWord,
    Refresh,
    Enter,
    Escape,
    Backspace,
    Quit,
    #[serde(skip)]
    Unhandled(char),
    Nop,
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn load_default_keymap() {
        let event_definitions: EventDefinitions = toml::from_str(DEFAULTS).unwrap();

        assert_debug_snapshot!(event_definitions, @"");
    }
}
