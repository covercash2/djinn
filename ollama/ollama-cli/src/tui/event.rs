use crossterm::event::{Event, KeyCode, KeyEvent};

#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub enum Action {
    Left,
    Right,
    Up,
    Down,
    LeftWord,
    RightWord,
    Refresh,
    Enter,
    Quit,
    #[default]
    Unhandled,
}

impl From<Event> for Action {
    fn from(event: Event) -> Self {
        match event {
            Event::FocusGained
            | Event::FocusLost
            | Event::Mouse(_)
            | Event::Paste(_)
            | Event::Resize(_, _) => Action::Unhandled,
            Event::Key(key_event) => key_event.into(),
        }
    }
}

impl From<KeyEvent> for Action {
    fn from(key_event: KeyEvent) -> Self {
        match key_event.code {
            KeyCode::Char('q') => Action::Quit,
            KeyCode::Char('j') => Action::Down,
            KeyCode::Char('k') => Action::Up,
            KeyCode::Up => Action::Up,
            KeyCode::Down => Action::Down,
            KeyCode::Char('h') => Action::Left,
            KeyCode::Char('l') => Action::Right,
            KeyCode::Char('r') => Action::Refresh,
            KeyCode::Enter => Action::Enter,
            _ => Action::Unhandled,
        }
    }
}
