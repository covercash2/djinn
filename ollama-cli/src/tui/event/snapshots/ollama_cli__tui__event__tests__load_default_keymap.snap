---
source: ollama-cli/src/tui/event/mod.rs
expression: event_definitions
snapshot_kind: text
---
EventDefinitions(
    {
        Edit: ActionMap(
            {
                Key {
                    event: KeyEvent {
                        code: Esc,
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Esc,
                        },
                    ),
                }: Escape,
                Key {
                    event: KeyEvent {
                        code: Down,
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Down,
                        },
                    ),
                }: Down,
                Key {
                    event: KeyEvent {
                        code: Enter,
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Enter,
                        },
                    ),
                }: Enter,
                Key {
                    event: KeyEvent {
                        code: Backspace,
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Backspace,
                        },
                    ),
                }: Backspace,
                Key {
                    event: KeyEvent {
                        code: Left,
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Left,
                        },
                    ),
                }: Left,
                Key {
                    event: KeyEvent {
                        code: Up,
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Up,
                        },
                    ),
                }: Up,
            },
        ),
        Normal: ActionMap(
            {
                Key {
                    event: KeyEvent {
                        code: Char(
                            '0',
                        ),
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Char(
                                '0',
                            ),
                        },
                    ),
                }: Beginning,
                Key {
                    event: KeyEvent {
                        code: Left,
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Left,
                        },
                    ),
                }: Left,
                Key {
                    event: KeyEvent {
                        code: Char(
                            'j',
                        ),
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Char(
                                'j',
                            ),
                        },
                    ),
                }: Down,
                Key {
                    event: KeyEvent {
                        code: Char(
                            'a',
                        ),
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Char(
                                'a',
                            ),
                        },
                    ),
                }: Edit,
                Key {
                    event: KeyEvent {
                        code: Char(
                            'w',
                        ),
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Char(
                                'w',
                            ),
                        },
                    ),
                }: RightWord,
                Key {
                    event: KeyEvent {
                        code: Char(
                            'i',
                        ),
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Char(
                                'i',
                            ),
                        },
                    ),
                }: Edit,
                Key {
                    event: KeyEvent {
                        code: Char(
                            'q',
                        ),
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Char(
                                'q',
                            ),
                        },
                    ),
                }: Quit,
                Key {
                    event: KeyEvent {
                        code: Char(
                            'h',
                        ),
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Char(
                                'h',
                            ),
                        },
                    ),
                }: Left,
                Key {
                    event: KeyEvent {
                        code: Up,
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Up,
                        },
                    ),
                }: Up,
                Key {
                    event: KeyEvent {
                        code: Char(
                            'k',
                        ),
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Char(
                                'k',
                            ),
                        },
                    ),
                }: Up,
                Key {
                    event: KeyEvent {
                        code: Esc,
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Esc,
                        },
                    ),
                }: Escape,
                Key {
                    event: KeyEvent {
                        code: Char(
                            'r',
                        ),
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Char(
                                'r',
                            ),
                        },
                    ),
                }: Refresh,
                Key {
                    event: KeyEvent {
                        code: Enter,
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Enter,
                        },
                    ),
                }: Enter,
                Key {
                    event: KeyEvent {
                        code: Backspace,
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Backspace,
                        },
                    ),
                }: Left,
                Key {
                    event: KeyEvent {
                        code: Down,
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Down,
                        },
                    ),
                }: Down,
                Key {
                    event: KeyEvent {
                        code: Char(
                            'b',
                        ),
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Char(
                                'b',
                            ),
                        },
                    ),
                }: LeftWord,
                Key {
                    event: KeyEvent {
                        code: Char(
                            '$',
                        ),
                        modifiers: KeyModifiers(
                            0x0,
                        ),
                        kind: Press,
                        state: KeyEventState(
                            0x0,
                        ),
                    },
                    node: Some(
                        Node {
                            modifiers: 0,
                            key: Char(
                                '$',
                            ),
                        },
                    ),
                }: End,
            },
        ),
    },
)
